### PR TITLE
New module postgresql_table - create postgresql table on remote host

### PR DESCRIPTION
**Pull request by Andersson007**
##### SUMMARY

Postgresql_table module allows to create (including LIKE statement), drop, rename, or truncate a table. Also allows to change some attributes, specifically, owner, tablespace, storage parameters.

##### ISSUE TYPE

- New Module Pull Request

##### MAIN PARAMETERS
```
db: databasename
table: tablename
state: present | absent
columns: col1 int primary key, col2 text not null, col3 biging
owner: owner
unlogged: yes | no
like: src_table
including: all
storage_params: fill_factor=100, autovacuum_vacuum_threshold=1
truncate: yes | no
rename: newname
```

##### EXAMPLES
```
- name: Create tbl2 in the acme database with the DDL like tbl1 with testuser as an owner
  postgresql_table:
    db: acme
    name: tbl2
    like: tbl1
    owner: testuser

- name: Create tbl2 in the acme database and tablespace ssd with the DDL like tbl1 including comments and indexes
  postgresql_table:
    db: acme
    table: tbl2
    like: tbl1
    including: comments, indexes
    tablespace: ssd

- name: Create test_table with several columns in ssd tablespace with fillfactor=10 and autovacuum_analyze_threshold=1
  postgresql_table:
    name: test_table
    columns: id bigserial primary key, num bigint, stories text
    tablespace: ssd
    storage_params: fillfactor=10, autovacuum_analyze_threshold=1

- name: Create an unlogged table
  postgresql_table:
    name: useless_data
    columns: waste_id int
    unlogged: true

- name: Rename table foo to bar
  postgresql_table:
    table: foo
    rename: bar

- name: Set owner to someuser
  postgresql_table:
    name: foo
    owner: someuser

- name: Change tablespace of foo table to new_tablespace and set owner to new_user
  postgresql_table:
    name: foo
    tablespace: new_tablespace
    owner: new_user

- name: Truncate table foo
  postgresql_table:
    name: foo
    truncate: yes

- name: Drop table foo
  postgresql_table:
    name: foo
    state: absent
```

##### RETURN
```
table:
  description: Name of a table.
  returned: always
  type: str
  sample: 'foo'
state:
  description: Table state.
  returned: always
  type: str
  sample: 'present'
owner:
  description: Table owner.
  returned: always
  type: str
  sample: 'postgres'
tablespace:
  description: Tablespace.
  returned: always
  type: str
  sample: 'ssd_tablespace'
query:
  description: Query that was tried to be execute.
  returned: always
  type: str
  sample: 'CREATE TABLE test_table (id bigint)'
storage_params:
  description: Storage parameters.
  returned: always
  type: list
  sample: [ "fillfactor=100", "autovacuum_analyze_threshold=1" ]
```

### CODE REVIEW

**Comment by Andersson007** [comment]
+label: postgresql

**Comment by Andersson007** [comment]
ready_for_review

**Comment by Andersson007** [comment]
cc @sebasmannem @amenonsen

**Comment by dagwieers** [comment]
@Andersson007 There is no need to 'cc' the same people again. We do this so these people are subscribed to the issue. As a result they will get every update to this ticket. Doing it again adds no value, except causes more mails to be send for no real value.

**Comment by Andersson007** [comment]
Ok, I didn't notice amenonsen was here.

**Comment by Andersson007** [comment]
@antoinell , I added the pg_qoute_identifier support here too ;)
I think it would be better you'll continue the great job.
We also have some ideas about new modules https://github.com/ansible/community/wiki/PostgreSQL.
If you don't have enough time for the new module development, support us by giving a valuable opinion or note.

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (Glandos) [inline lib/ansible/modules/database/postgresql/postgresql_table.py:560]: You can use [`ensure_libs`](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/postgres.py#L42).

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (Glandos) [inline lib/ansible/modules/database/postgresql/postgresql_table.py:568]: And remove this afterwards.

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:560]: @dagwieers , could you give us your opinion ?

Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:560]: I need opinions of experienced contributors, @gundalow  , maybe you know, is this common way to check libs now?

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:568]: I can't do it without the approval by older brothers. Thank you for the comment

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (tcraxs) [inline lib/ansible/modules/database/postgresql/postgresql_table.py:572]: In addition to @Glandos comment, I think you can use it like so (or see [here](https://github.com/Glandos/ansible/blob/f4204aa69c7e3429b83b3d5d7d5f7ee07f29407e/lib/ansible/modules/database/postgresql/postgresql_db.py#L451)):
```python
import ansible.module_utils.postgres as pgutils
...
    try:
        pgutils.ensure_libs(sslrootcert=module.params.get('ssl_rootcert'))
        db_connection = psycopg2.connect(**kw)
...
    except pgutils.LibraryError as e:
         module.fail_json(msg="unable to connect to database: {0}".format(to_native(e)), exception=traceback.format_exc())
    except TypeError as e:
        if 'sslrootcert' in e.args[0]:
            module.fail_json(msg='Postgresql server must be at least version 8.4 to support sslrootcert',
                exception=traceback.format_exc())
...
```

You could also add `exception=traceback.format_exc()`to your exceptions like [here](https://github.com/Glandos/ansible/blob/f4204aa69c7e3429b83b3d5d7d5f7ee07f29407e/lib/ansible/modules/database/postgresql/postgresql_db.py#L462)

Comment (tcraxs) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:572]: and then remove the parts Glandos point you.

**Review by tcraxs — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:572]: @tcraxs , thank you for this note!
I tried to change it.
However, look at lines 134, 422, 451 [here](https://github.com/Glandos/ansible/blob/f4204aa69c7e3429b83b3d5d7d5f7ee07f29407e/lib/ansible/modules/database/postgresql/postgresql_db.py#L451) (your first link)

1. postgresql_db.py imports psycopg2 in line 134
2. anyway, it checks that psycopg2 was successfully imported in line 422
3. ensure_libs checks again that psycopg2 was imported and its version in addition.

Anyway, we must import psycopg2 in the module file and if psycopg2.version < '2.4.3' and sslrootcert is not None: is a quite simple condition to import and invoke a special function ensure_libs to avoid it.

To be short, I suggest to leave the current version of this module related with psycopg2 and its version as it is because it's more efficient and simplier.

**Comment by Andersson007** [comment]
ready_for_review

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (tcraxs) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:572]: I'm totally fine with your explanation and it was just a starting point by my side. And you are right, that the usage in your module in this case is simpler, so no change needed.

**Comment by tcraxs** [comment]
@Andersson007 Good work. New module LGFM, tests are also working fine for me.

**Comment by tcraxs** [comment]
shipit

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_table.py]:
```suggestion
```

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_table.py]:
```suggestion
version_added: '2.8'
```

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_table.py]: Needs `Postgresql` to appear in here somewhere, and maybe shortening a bit
Maybe change to `Create or modify Postgresql table attributes`

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (gundalow) [reply lib/ansible/modules/database/postgresql/postgresql_table.py:560]: A follow up PR that updates all postgres modules to use https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/postgres.py#L42 would be good. No need to block on this.

Also that function should be updated to use `module.fail_json(msg=missing_required_lib("psycopg2"))`

File: lib/ansible/modules/database/postgresql/postgresql_table.py
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_table.py]: Yes, I agree, it looks quite complex now

**Comment by Andersson007** [comment]
added module.fail_json(msg=missing_required_lib("psycopg2"))

**Comment by gundalow** [comment]
`module_utils/postgres.py` defines `ca_cert`

**Comment by Andersson007** [comment]
Ok, thx! I'll solve it later

**Comment by Andersson007** [comment]
fixed

**Comment by gundalow** [comment]
@Andersson007 @tcraxs Thanks. Merged into `devel` for release in Ansible 2.8

**Comment by Andersson007** [comment]
thanks all !

**Comment by frinux** [comment]
If this module aims also to drop tables (as the state=absent is possible), it should be great to add a 'cascade' option. In most of the case, the DROP will fail due to foreign key constraints.

**Comment by Andersson007** [comment]
@frinux , hello, thank you for your feedback! I’m on a public holiday before Monday. Will try to look on the next week

**Comment by Andersson007** [comment]
@frinux , very good suggestion! thank you, done #56068 . Would be better if you look at it and aprove

**Pull request merged**
